### ISSUE TITLE
Binary diff showing /dev/null

### ISSUE THREAD

**Issue by kevinjswinton** [libgit2/libgit2#5493]
### Reproduction steps

1. Use the "create_repo.sh" file as a BASH script to construct a minimal repo for testing purposes.
2. Then compile the "main.c" client to link with libgit2 and demonstrate the behaviour on the minimal repo.

[files.zip](https://github.com/libgit2/libgit2/files/4496495/files.zip)

### Expected behavior

<pre>
diff --git a/ascfile b/ascfile
index 46e25a7..249c1f3 100644
--- a/ascfile
+++ b/ascfile
@@ -1 +1 @@
-Hello World 1
+Hello World 2!
diff --git a/binfile b/binfile
index ff08ca1..<strong>f43e1ec</strong> 100755
<strong>Binary files a/binfile and b/binfile differ</strong>
</pre>

### Actual behavior

<pre>
diff --git a/ascfile b/ascfile
index 46e25a7..249c1f3 100644
--- a/ascfile
+++ b/ascfile
@@ -1 +1 @@
Hello World 1
Hello World 2!
diff --git a/binfile b/binfile
index ff08ca1..<strong>0000000</strong> 100755
<strong>Binary files a/binfile and /dev/null differ</strong>
</pre>

### Version of libgit2 (release number or SHA1)

1.0 (commit 918a7d19553a9a4181bf1561b8b828a5f82e60e0)

### Operating system(s) tested

Debian 64-bit, RISC OS 5.27

**Comment by kevinjswinton** [comment]
I have already dug into this and resolved it. I will submit a Pull Request shortly.

**Comment by pks-t** [comment]
Thanks a lot for your bug report! I'm afraid though that this is not a bug, but expected behaviour: by default, we do not generate binary diffs. To fix this, you'll have to change your code like follows:

```c
#include <stdio.h>
#include "git2.h"

int cli_diff_printer(  const git_diff_delta *delta, const git_diff_hunk *hunk, const git_diff_line *line, void *payload ) {

        printf( "%.*s", line->content_len, line->content );
        return 0;

}

int main( int argc, char **argv ) {
        git_diff_options opts = GIT_DIFF_OPTIONS_INIT;
        git_repository *repo;
        git_diff *diff;
        void *payload;

        /* Instruct the diff mechanism to generate binary patches. */
        opts.flags = GIT_DIFF_SHOW_BINARY;

        git_libgit2_init();
        git_repository_open( &repo, ".git" );
        git_diff_index_to_workdir( &diff, repo, NULL, &opts );
        git_diff_print( diff, GIT_DIFF_FORMAT_PATCH, cli_diff_printer, payload );

        return 0;

}
```

Most importantly, you need to set up the `git_diff_options` structure and set its flag field to include `GIT_DIFF_SHOW_BINARY`. You then pass the struct to `git_diff_ndex_to_workdir` as its last argument.

Please let me know if this fixes the bug for you.

**Comment by kevinjswinton** [comment]
Hi, and thanks for the quick response - much appreciated (especially during current times).

By dropping in the code suggested, although ``GIT_DIFF_SHOW_BINARY`` corrects the missing OID, the summary line:

```Binary files a/binfile and b/binfile differ```

is replaced with binary delta output:

```
GIT binary patch
literal 198976
zc%1Bgd0bRw^!L33qp0A;1jEuo8x3*^Op;M2K|$}>P?TC$7AlHTs3bFr<$}(PavR5}
zw9Gc!ZMTigTv!y;7Pl62*HZ5lca2S9-gC}#XSg#tzxVfk-hac7+<U(F+~+yxdCqg5
...
```

To illustrate, having run the amended client code suggested, libgit2 now outputs this:

```
diff --git a/ascfile b/ascfile
index 46e25a7..249c1f3 100644
--- a/ascfile
+++ b/ascfile
@@ -1 +1 @@
Hello World 1
Hello World 2!
diff --git a/binfile b/binfile
index ff08ca1da103b8be5a62a9822d5ea112dedab8d8..f43e1ecf13751ad999e29d53da103352e97d84c5 100755
GIT binary patch
literal 198976
zc%1Bgd0bRw^!L33qp0A;1jEuo8x3*^Op;M2K|$}>P?TC$7AlHTs3bFr<$}(PavR5}
zw9Gc!ZMTigTv!y;7Pl62*HZ5lca2S9-gC}#XSg#tzxVfk-hac7+<U(F+~+yxdCqg5
...
```

This looks closer to the ``git diff --text`` command than the more basic ``git diff`` command (which is what I was hopeful for). Is there something different I should be doing that will allow for libgit2 to match more successfully the behaviour/output of the basic "git diff" command (where libgit2 doesn't lose sight of the OID and pathname?)

**Comment by pks-t** [comment]
Sorry for dropping the ball here. I tried to come up with another testcase that reproduces the error, but failed:

```c
void test_diff_binary__binary_null(void)
{
	git_buf patch = GIT_BUF_INIT;
	git_index *index;
	git_diff *diff;

	repo = cl_git_sandbox_init("empty_standard_repo");

	cl_git_write2file("empty_standard_repo/blob.ascii", "Hello world 1\n", 14,
			  O_CREAT|O_EXCL|O_WRONLY, 0644);
	cl_git_write2file("empty_standard_repo/blob.bin", "binary\0content\0", 15,
			  O_CREAT|O_EXCL|O_WRONLY, 0644);
	cl_git_pass(git_repository_index(&index, repo));
	cl_git_pass(git_index_add_bypath(index, "blob.ascii"));
	cl_git_pass(git_index_add_bypath(index, "blob.bin"));
	cl_repo_commit_from_index(NULL, repo, NULL, 1372350000, "Base");

	cl_git_write2file("empty_standard_repo/blob.ascii", "Hello world 2\n", 14,
			  O_TRUNC|O_WRONLY, 0644);
	cl_git_write2file("empty_standard_repo/blob.bin", "binary\0xontent\0", 15,
			  O_TRUNC|O_WRONLY, 0644);

	cl_git_pass(git_diff_index_to_workdir(&diff, repo, NULL, NULL));
	cl_git_pass(git_diff_to_buf(&patch, diff, GIT_DIFF_FORMAT_PATCH));
	cl_assert_equal_s(patch.ptr,
		"diff --git a/blob.ascii b/blob.ascii\n"
		"index 560fefb..b64ccf4 100644\n"
		"--- a/blob.ascii\n"
		"+++ b/blob.ascii\n"
		"@@ -1 +1 @@\n"
		"-Hello world 1\n"
		"+Hello world 2\n"
		"diff --git a/blob.bin b/blob.bin\n"
		"index 9886fb0..6c097a6 100644\n"
		"Binary files a/blob.bin and b/blob.bin differ\n");

	git_buf_dispose(&patch);
	git_index_free(index);
	git_diff_free(diff);
}
```

The output seems to match what you want, right? Am I doing anything differently compared to your script that you can spot?

**Comment by kevinjswinton** [comment]
Hi Patrick

Now this gets interesting - I've just cloned a fresh copy of libgit2. I've built it, linked it against the original main.c I supplied (where we don't set GIT_DIFF_SHOW_BINARY) and there's something I've just spotted that might be of interest:

When I make use of the example you just supplied, everything works and the output is what we would expect ("Binary files a/blob.bin and b/blob.bin differ"). So that makes sense: your test case, and my implementation of your test case, look fine.

But. If I run the code against the repository when I'm using `/usr/bin/ls` and `/usr/bin/grep` as my example binary files, then it fails, and the output I'm getting is the output I was originally showing in this thread.

So. I've cloned the latest copy of libgit2. I've built against that. I'm still using the main.c example (without setting GIT_DIFF_SHOW_BINARY as you previously suggested). And the only obvious difference here - from what I can see - suggests it may relate to the _size/content_ of the binary files: If I use your simple examples where you just pipe in "binary\0content\0" and then "binary\0xontent", these run for me as they do for you. But if I drop them in favour of larger and "more serious" binary files (such as /usr/bin/ls and /usr/bin/grep), then the problem I originally explained pops up and I still see libgit2's diff output stating:

``
Binary files a/binfile and /dev/null differ
``

Is there perhaps some kind of file size limit kicking in here, that might be surfacing this issue? (Note I'm still running this on a stock Debian VM I've fired up). If you pop in the usage of /usr/bin/ls and /usr/bin/grep for your binary files, I wonder if you might now be able to reproduce what I'm seeing here?

Thanks for your time on this - appreciated as always!
